### DISCUSSION TITLE
Autoryzacja i konfiguracja aplikacji webowej przy wykorzystaniu Azure AD #12

### DISCUSSION THREAD

**Discussion by odzio33** [AzureCommunityPL/group#134] [Pomoc] [answered]
Posiadam aplikacje webowa, została ona wrzucona do Azura - ustawiony connector itd. Generalnie robimy przejście na sharepoint, który będzie głównym narzędziem pracy, z sharepointa będą odnośniki do naszej aplikacji i tutaj musimy zautoryzować użytkownika czy jest zalogowany czy nie i go puścić. Jaki sposób najlepiej wybrać? Jest SAML, oAutch2, Connect ID, nagłówki i pewnie wile narzędzie jakie udostępnia Microsoft + ewentualnie obsługa na apachu poprzez moduły.

Jaki protokół wybrać?

Napisana została testowa aplikacja w java: hello word z filtrem która robi autentykacje za pomocą biblioteki MSAL i faktycznie sprawdza czy użytkownik jest zalogowany czy nie i puszcza nas na stronę lub redirect do logowania microsoftu. Jak później badać czy użytkownik jest zalogowany? Czy co reqest trzeba robić hint do microsoftu?

**Comment by piotrgwiazda** [comment]
OAuth2 wybieramy w 2020 roku.

Jeśli MSAL w JS to tylko wersja 2.0 w modelu PKCE. Flow Implicit (w wersji
MSAL 1) jest przestarzały od 10 lat. Wtedy serwer tylko sprawdza token JWT
jakąś biblioteką, ale sam nie robi procesu uwierzytelniania. Tu trzeba
uważać na przechowywanie tokena.

Jeśli uwierzytelnianie po stronie serwera to Auth Code Flow. Wtedy masz
normalną sesję z ciastkiem itp., JS nic nie robi, całość dzieje się po
stronie serwera.

Jak stworzysz w AAD nowy App Registration to samouczki i quick starty
poprowadzą Cię przez to.

śr., 4 lis 2020, 13:58 użytkownik odzio33 <notifications@github.com>
napisał:

**Reply by odzio33** [reply]
A umiałbyś przedstawić samo story jak to ma działać - bez kodowania? 
Ja przechodziłem przez te samouczki, inaczej nie szukałbym pomocy na forach :)

Krok pierwszy jest taki że strzelam w endpoint /authorize

![image](https://user-images.githubusercontent.com/22906225/98236445-a5464000-1f63-11eb-9460-ee3b42875a29.png)

I co dalej? W odpowiedzi mam userId (email zalogowanego lub przekierowanie na logowanie microsoft)
Jak to powinno działać, przpływ informacji, sesja.
Co request trzeba weryfikować zalogowanie?

**Comment by mgrabarz** [comment]
@odzio33 ja mam wrażenie że sporo przed tobą. Nawet nazw tych wszystkich "sposobów" nie wymieniłeś poprawnie.

Jeśli pozostaniesz z logowaniem poprzez Azure AD, a zakładam że to jest stan obecny z przekierowaniami, to sposób sprawdzenia czy ktoś jest zalogowany sprowadza się do nagłówka w żądaniu http.

```
Authorization: bearer tutaj-twoj-jwt
```

Token JWT trzeba sprawdzić pod kątem: iss, sub, aud, exp, nbf i podpisu. Więcej na ten temat bez problemu znajdziesz online.

**Comment by piotrgwiazda** [comment]
Auth0 ma ładne obrazkowe opisy Oauth2 Flows. https://auth0.com/docs/flows
Polecam

czw., 5 lis 2020, 12:42 użytkownik odzio33 <notifications@github.com>
napisał:

**Comment by odzio33** [comment]
Dziękuje za odpowiedzi (@piotrgwiazda, @mgrabarz), ale wydaję mi się że mój problem wygląda nieco inaczej. Dla sprostowania opiszę go nieco dokładniej.

Potrzebuje wsparcia/konsultacji w zakresie opracowania rozwiązania SSO, które pozwoli na pełną integrację sesji użytkownika pomiędzy moją aplikacją (SZOP)  a  Sharepoint. 

Wymagania w tym obszarze są następujące:
•	Logując się do Sharepoint powinniśmy być równocześnie zalogowani do SZOP (i odwrotnie
logując się do SZOP powinniśmy być równocześnie zalogowani do Sharepoint). 
•	Wylogowanie powinno działać analogicznie .
•	Jeżeli niezalogowany użytkownik wpisze w przeglądarce adres http chronionego zasobu, powinien zostać przekierowany na  ekran logowania i po pomyślnym uwierzytelnieniu na chroniony zasób.

Na chwilę obecną moje aplikacje są wystawione przez Azure Proxy pod adresem xyz.pl/xxx/ i dostęp do nich wymaga wcześniejszego zalogowania do Sharepoint (pre-autentykacja Azure).  Są to aplikacje „java web application” uruchomione na serwerze Tomcat.
Jak się wydaje głównym problemem jest to, że na poziomie moich aplikacji (SZOP) nie potrafię zidentyfikować zalogowanego użytkownika. 

Analizowałem możliwość wykorzystanie standardu OpenID, którego implementacja jest dostępna w Azure. Widzę tu jednak pewne zagrożenie, które przedstawia scenariusz przedstawiony poniżej.
•	Użytkownik loguje się do Sharepoint-a 
•	Użytkownik klika w link do naszej aplikacji wewnętrznej
•	Aplikacja (SZOP) stwierdza że użytkownik nie jest zalogowany, następnie wykorzystując standard OpenId identyfikuje go i  tworzy dla niego sesję.
•	Użytkownik może korzystać z SZOP w prawidłowy sposób
•	Użytkownik wylogowywuje się z Sharepoint-a
•	Nie możliwości uruchomienia żadnej aplikacji wewnętrznej

Do tej pory wszystko wydaje się działać zgodnie z założeniami. Problem pojawi się wg nas wtedy, gdy inny użytkownik na tym samym komputerze zaloguje się do Sharepointa, wtedy sesja dla aplikacji SZOP będzie już istniała jednak będzie to sesja poprzednio zalogowanego użytkownika.

Pozdrawiam,
PO

**Comment by piotrgwiazda** [comment, marked as answer]
Twój problem polega zatem na poprawnej obsłudze sesji/stanu/ciastek.
Ponownie odsyłam dokumentacji.

Azure AD trzyma sesję (przycisk "Nie wylogowuj mnie" czy jakoś tak) i jak
użytkownik się zaloguje w Azure AD to wchodzi na Sharepoint, Office, Teams
i na Twoją aplikację bez ponownego uwierzytelniania. Każda aplikacja
otrzymuje tokeny ID token i Access token dla siebie (pole audience) i z
jakimś czasem ważności, np godzinę. Zakładamy, że przez godzinę użytkownik
nie straci pracy ;-)

W Tomcacie Ty decydujesz. Jeśli przestajesz bazować na tokenie po tym jak
zautorysujesz użytkownika i polegasz na ciastku sesji, to musisz sprawdzić
jak to cistko "działa". Upewnij się np. że wygasa z chwilą zamknięcia
przeglądarki. To już nie ma zwiazku z AAD.



czw., 12 lis 2020, 11:38 użytkownik odzio33 <notifications@github.com>
napisał:

**Comment by mgrabarz** [comment]
W zasadzie zgadzam się z diagnozą @piotrgwiazda, przy czym dochodzi jeszcze poprawna implementacja wylogowania w aplikacji shop.

Na potrzeby shopa określ jak przechowujesz sesje i czym ona jest. To pewnie ułatwi ci złożenie tego w całość. Kolejny aspekt to SessionStorage vs LocalStorage.
